packages/yarnpkg-pnpify/sources/RawHoister.ts

```diff
@@ -0,0 +1,244 @@
+ * package, it will be searched over the folder of its parent package dependencies. And
+ * hence we can lift dependencies that have different names to the parent package "dependencies",
+ * thus reducing tree height.
+ *
+ * Theoretical computational complexity of hoisting algorhitm is O(n^2) (when no hoisting can be
+ * done on all the tree), but in practice it should make much less iterations over tree, due to
+ * real dependency trees more often then not can be almost fully hoisted.
+ */
+export class RawHoister {
+  /**
+   * Hoists package tree, by applying hoisting algorithm to each tree node independently.
+   * We first try to hoist all the packages from anywhere in the whole tree to the root package.
+   * Then we apply the same algorithm to the subtree that starts at one of the dependencies of the
+   * root package, we do this for each dependency, then move further down to dependencies of
+   * dependencies, etc.
+   *
+   * @param tree package tree
+   * @param packageMap package map
+   * @param nohoist package ids that should be excluded from applying hoisting algorithm. Nohoist
+   *                packages can be hoisted themselves, and their dependencies can be hoisted too,
+   *                but only to the package itself, they cannot be hoisted to the nohoist package parent
+   */
+  public hoist(tree: PackageTree, packageMap: PackageMap, nohoist: Set<PackageId> = new Set()): PackageTree {
+    // Make normalized tree copy, which will be mutated by hoisting algorithm
+    const treeCopy = this.cloneTree(tree);
+
+    const hoistSubTree = (nodeId: PackageId) => {
+      // Apply mutating hoisting algorithm on each tree node starting from the root
+      this.hoistInplace(treeCopy, nodeId, packageMap, nohoist);
+
+      for (const depId of treeCopy.get(nodeId) || NO_DEPS) {
+        hoistSubTree(depId);
+      }
+    };
+
+    hoistSubTree(0);
+
+    return treeCopy;
+  }
+
+  /**
+   * Performs package subtree hoisting to its root.
+   * This funtion mutates tree.
+   *
+   * @param tree package tree
+   * @param rootId package subtree root package id
+   * @param packageMap package map
+   * @param nohoist nohoist package ids
+   */
+  private hoistInplace(tree: PackageTree, rootId: PackageId, packageMap: PackageMap, nohoist: Set<PackageId>): void {
+    // Get the list of package ids that can and should be hoisted to the subtree root
+    const hoistedDepIds = this.computeHoistCandidates(tree, rootId, packageMap, nohoist);
+
+    const removeHoistedDeps = (nodeId: PackageId) => {
+      // No need to traverse past nohoist node
+      if (nohoist.has(nodeId))
+        return;
+
+      const depIds = tree.get(nodeId) || NO_DEPS;
+      for (const depId of depIds) {
+        // First traverse to deeper levels
+        removeHoistedDeps(depId);
+        // Then remove hoisted deps from current node
+        if (hoistedDepIds.has(depId)) {
+          depIds.delete(depId);
+        }
+      }
+      // Remove node without deps
```

Comment:
Why? Packages without deps seem likely to exist, no? 🤔

Reply:
Yes, they do. But why should we keep an entry for them in the tree, its a waste of memory

Reply:
Ah yeah, I forgot you were doing a `|| NO_DEPS` to account for those "empty spaces" 👍